common/ast/src/parser/ast/expression.rs

```diff
@@ -22,73 +22,68 @@ use crate::parser::ast::display_identifier_vec;
 #[derive(Debug, Clone, PartialEq)]
 #[allow(dead_code)]
 pub enum Expr {
-    // Wildcard star
-    Wildcard,
-    // Column reference, with indirection like `table.column`
+    /// Column reference, with indirection like `table.column`
     ColumnRef {
         database: Option<Identifier>,
         table: Option<Identifier>,
         column: Identifier,
     },
-    // `IS NULL` expression
-    IsNull(Box<Expr>),
-    // `IS NOT NULL` expression
-    IsNotNull(Box<Expr>),
-    // `[ NOT ] IN (expr, ...)`
+    /// `IS [ NOT ] NULL` expression
+    IsNull { expr: Box<Expr>, not: bool },
+    /// `[ NOT ] IN (expr, ...)`
     InList {
         expr: Box<Expr>,
         list: Vec<Expr>,
         not: bool,
     },
-    // `[ NOT ] IN (SELECT ...)`
+    /// `[ NOT ] IN (SELECT ...)`
     InSubquery {
         expr: Box<Expr>,
         subquery: Box<Query>,
         not: bool,
     },
-    // `BETWEEN ... AND ...`
+    /// `BETWEEN ... AND ...`
     Between {
         expr: Box<Expr>,
-        negated: bool,
         low: Box<Expr>,
         high: Box<Expr>,
+        not: bool,
     },
-    // Binary operation
+    /// Binary operation
     BinaryOp {
         op: BinaryOperator,
         left: Box<Expr>,
         right: Box<Expr>,
     },
-    // Unary operation
-    UnaryOp {
-        op: UnaryOperator,
-        expr: Box<Expr>,
-    },
-    // `CAST` expression, like `CAST(expr AS target_type)`
+    /// Unary operation
+    UnaryOp { op: UnaryOperator, expr: Box<Expr> },
+    /// `CAST` expression, like `CAST(expr AS target_type)`
     Cast {
         expr: Box<Expr>,
         target_type: TypeName,
     },
-    // A literal value, such as string, number, date or NULL
+    /// A literal value, such as string, number, date or NULL
     Literal(Literal),
-    // Scalar function call
+    /// `COUNT(*)` expression
+    CountAll,
+    /// Scalar function call
     FunctionCall {
-        // Set to true if the function is aggregate function with `DISTINCT`, like `COUNT(DISTINCT a)`
+        /// Set to true if the function is aggregate function with `DISTINCT`, like `COUNT(DISTINCT a)`
         distinct: bool,
         name: String,
         args: Vec<Expr>,
         params: Vec<Literal>,
     },
```

Comment:
I'm thinking about lifting the `distinct` field out and add a `AggregationFunc` variant for aggregation functions(SUM, AVG etc.).

Reply:
In the level of `AST`, can we distinguish `functions` and `aggregate functions` ?

Reply:
@sundy-li I think it's viable because the name of aggregate functions are keywords.

Reply:
Yes. But if we have custom aggregate functions like: `window_funnel`, `retention`, should we add them into 
 as keywords?

Reply:
> Yes. But if we have custom aggregate functions like: `window_funnel`, `retention`, should we add them into as keywords?

In general, it's case by case, but my suggestion is treating aggregate functions names as keywords.

It's more convenient to handle special grammar of specific functions by reserving their names as keyword(e.g. `EXTRACT(... FROM ...)`, `COUNT(*)`, `SUM(DISTINCT ...)`).

Since all aggregate functions should support the `DISTINCT` grammar, and the number of aggregate functions is much lower than scalar functions, I think it's fine to reserve their names as keywords.

Reply:
> But there are two main problems I came up with:
> 
> 1. UDF Aggregate/Scalar Functions （function name is defined by the user  and stored in meta service)
> 2. Combinators:  `*If`  like sumIf.    If we add some aggregate function, we should update the keywords & combinator keywords.

The `UDF`s are context sensitive that cannot be handled by parser.

It seems we have to keep the `distinct` field here for planner to check the semantic of a unrecognizable function...

What about only reserving the special cases like `COUNT`(for `COUNT(*)`) as keyword?